python/grass/script/tests/utils_test.py

```diff
@@ -21,3 +23,28 @@ def test_backslash_separators():
 def test_unrecognized_separator():
     """Check that unknown strings are just passed through"""
     assert gs.separator("apple") == "apple"
+
+
+def test_KeyValue():
+    """Check that KeyValue class works as expected"""
+    kv = gs.KeyValue()
+    kv["key1"] = "value1"
+    kv["key2"] = "value2"
+
+    assert kv["key1"] == "value1"
+    assert kv.key1 == "value1"
+    assert kv["key2"] == "value2"
+    assert kv.key2 == "value2"
+
+    # Check that it raises KeyError for non-existing keys
+    with pytest.raises(KeyError):
+        _ = kv["non_existing_key"]
+
+    # Check that it raises AttributeError for non-existing attribute
+    with pytest.raises(AttributeError):
+        _ = kv.non_existing_key
+
+    # Deep copy works correctly
+    from copy import deepcopy
+
+    _ = deepcopy(gs.KeyValue())
```

Comment:
Is there anything else to check for in this? There's no assertion on anything. 
I understand that for now, it allows to check that there's no failure.

Reply:
No, it’s not noise, it’s rather that maybe it isn’t strong enough to let us catch if it doesn’t work later on. 

But on second thought, we don’t have to test for the functionality of the library function deepcopy.

But we might want to actually have something to copy for, that isn’t immutable, and that requires to go recursivly. Something like a list for one (that an element can be changed after the copy), and an object of a class, it could even be another KeyValue as a value, in order to know it goes down a level.

Reply:
Also, in that test function, you essentially have 4 different unit tests that could be separated. (So a failure means that one single spec/feature is not working).